core/cloudflow-operator/src/main/scala/cloudflow/operator/event/ConfigInputChangeEvent.scala

```diff
@@ -0,0 +1,259 @@
+   */
+  def fromWatchEvent()(implicit system: ActorSystem): Flow[WatchEvent[Secret], ConfigInputChangeEvent, NotUsed] =
+    Flow[WatchEvent[Secret]]
+      .statefulMapConcat { () ⇒
+        var currentSecrets = Map[String, WatchEvent[Secret]]()
+        watchEvent ⇒ {
+          val secret       = watchEvent._object
+          val metadata     = secret.metadata
+          val secretName   = secret.metadata.name
+          val namespace    = secret.metadata.namespace
+          val absoluteName = s"$namespace.$secretName"
+          val data         = getData(secret)
+
+          watchEvent._type match {
+            case EventType.DELETED ⇒
+              currentSecrets = currentSecrets - absoluteName
+              List()
+            case EventType.ADDED | EventType.MODIFIED ⇒
+              if (currentSecrets.get(absoluteName).forall { existingEvent ⇒
+                    // the secret must have been updated
+                    existingEvent._object.resourceVersion != watchEvent._object.resourceVersion &&
+                    // the secret must change
+                    getData(existingEvent._object) != data
+                  }) {
+                currentSecrets = currentSecrets + (absoluteName -> watchEvent)
+                (for {
+                  appId        ← metadata.labels.get(Operator.AppIdLabel)
+                  configFormat <- metadata.labels.get(CloudflowLabels.ConfigFormat) if configFormat == CloudflowLabels.InputConfig
+                  _ = system.log.info(s"[app: $appId configuration changed ${changeInfo(watchEvent)}]")
+                } yield {
+                  ConfigInputChangeEvent(appId, namespace, watchEvent)
+                }).toList
+              } else List()
+          }
+        }
+      }
+
+  def getData(secret: Secret): String =
+    secret.data.get(ConfigInputChangeEvent.SecretDataKey).map(bytes => new String(bytes, StandardCharsets.UTF_8)).getOrElse("")
+
+  private def changeInfo[T <: ObjectResource](watchEvent: WatchEvent[T]) = {
+    val obj      = watchEvent._object
+    val metadata = obj.metadata
+    s"(${getKind(obj)} ${metadata.name} ${watchEvent._type})"
+  }
+
+  private def getKind(obj: ObjectResource) = if (obj.kind.isEmpty) obj.getClass.getSimpleName else obj.kind // sometimes kind is empty.
+
+  /**
+   * Finds the associated [[CloudflowApplication.CR]]s for [[ConfigInputChangeEvent]]s.
+   * The resulting flow outputs tuples of the app and the streamlet change event.
+   */
+  def mapToAppInSameNamespace(
+      client: KubernetesClient
+  )(implicit ec: ExecutionContext): Flow[ConfigInputChangeEvent, (Option[CloudflowApplication.CR], ConfigInputChangeEvent), NotUsed] =
+    Flow[ConfigInputChangeEvent].mapAsync(1) { configInputChangeEvent ⇒
+      val ns = configInputChangeEvent.watchEvent._object.metadata.namespace
+      client.usingNamespace(ns).getOption[CloudflowApplication.CR](configInputChangeEvent.appId).map {
+        case a @ Some(_) ⇒ a    -> configInputChangeEvent
```

Comment:
weird construction. It doesn't look like it's doing something with the value. I think it can be simplified to:
```
.map{cr => cr -> configInputChangeEvent
```
or, if possible,
```
configInputChangeEvent -> cr
```
I'm not a fan of having a `None` as a key.

Reply:
After seeing `toInputConfigUpdateAction`, I think you can safely `flatMap` the value and get rid of the `None` case

Reply:
I have to emit an element in the flow when it is not there.

Reply:
Ah I see what you mean.. must have been different before and not fully cleaned up.

Reply:
@maasg yeah there is a little copy/paste going on here.. the mapToAppInSameNamespace is used slightly differently in StreamletChangeEvents. will cleanup.